[New-45/47/49] - Articles, Categories, Tags business logic implementation.

EvgeniyRuban opened the pull request.

EvgeniyRuban committed:
Articles repository implemeted.

EvgeniyRuban committed:
Added Article service.

EvgeniyRuban committed:
Added Article controller.

EvgeniyRuban committed:
Added Category and Tag repositories.

EvgeniyRuban committed:
Added Category and Tag controllers.

EvgeniyRuban committed:
Tag and Category controllers updated.

EvgeniyRuban committed:
Updated Article service.

EvgeniyRuban committed:
Added summary comments for the contracts.

EvgeniyRuban committed:
Added summary comments for the contracts.

EvgeniyRuban committed:
Merge branch 'articles-businesslogic-impl' of https://github.com/GBDev1-FirstTeam/new-habr into articles-businesslogic-impl

Narendin commented on 01. Presentation/NewHabr.WebApi/Controllers/ArticleController.cs:
```diff
@@ -0,0 +1,155 @@
+
+    [HttpPost]
+    public async Task<ActionResult> CreateAsync([FromQuery] CreateArticleRequest request, CancellationToken cancellationToken = default)
+    {
```
а чего FromQuery, а не FromBody?

Narendin commented on 01. Presentation/NewHabr.WebApi/Controllers/TagController.cs:
```diff
@@ -0,0 +1,110 @@
+    public async Task<ActionResult> UpdateAsync([FromQuery] TagDto tagToUpdate, CancellationToken cancellationToken = default)
+    {
+        if (tagToUpdate is null)
+        {
```
эти проверки надо выносить в IValidatable, либо в отдельные валидаторы

Narendin commented on 02. Business/NewHabr.Business/NewHabr.Business.csproj:
```diff
@@ -7,11 +7,10 @@
   <ItemGroup>
-    <ProjectReference Include="..\NewHabr.Domain\NewHabr.Domain.csproj" />
+    <PackageReference Include="AutoMapper" Version="12.0.1" />
   </ItemGroup>
```
Автомаппер добавлен в NewHabr.Domain.csproj
Ссылка на него есть
Предгалаю отседова этот пакет выпилить

EvgeniyRuban replied:
Я его добавил потому, что описывал ошибки в саммори комментах для методов интерфейсов.

Narendin commented on 02. Business/NewHabr.Domain/Contracts/IArticleService.cs:
```diff
@@ -0,0 +1,43 @@
+using AutoMapper;
+
+namespace NewHabr.Domain.Contracts;
+
```
Может, интерфейсы репозиториев и сервисов в отдельные папки раскидаем?

EvgeniyRuban replied:
Согласен

Narendin commented on 03. DAL/NewHabr.DAL/Repository/ArticleRepository.cs:
```diff
@@ -1,33 +1,45 @@
+    public async Task DeleteByIdAsync(Guid id, CancellationToken cancellationToken = default)
     {
-        throw new NotImplementedException();
+        var article = await Set.FirstOrDefaultAsync(a => a.Id == id && !a.Deleted, cancellationToken);
```
вот для таких случаев в базовый репозиторий надо добавить метод GetById()!

Narendin commented on 03. DAL/NewHabr.DAL/Repository/ArticleRepository.cs:
```diff
@@ -1,33 +1,45 @@
 
-    public Task<IReadOnlyCollection<Article>> GetDeletedAsync(CancellationToken cancellationToken = default)
+    public async Task<IReadOnlyCollection<Article>> GetDeletedAsync(CancellationToken cancellationToken = default)
     {
```
Метод GetDeletedAsync так же можно смело выносить в базовый репозиторий. Реализация его не отличается нигде, все сущности имеют ключ Deleted, зачем дублировать?

Narendin commented on 03. DAL/NewHabr.DAL/Repository/TagRepository.cs:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    public async Task DeleteByIdAsync(int id, CancellationToken cancellationToken = default)
+    {
```
ровно то же самое. В Базовый репозиторий. Ничем особенным эта реализация не выделяется

SimeonIksanov commented on 01. Presentation/NewHabr.WebApi/Controllers/ArticleController.cs:
```diff
@@ -0,0 +1,155 @@
+
+[Route("[controller]")]
+[ApiController]
+public class ArticleController : Controller
```
при написании API рекомендуется наследоваться от ControllerBase

SimeonIksanov replied:
Учитывая что имя контроллера появится в маршруте, и учитывая рекомендации https://restfulapi.net/resource-naming/ предлагаю переименовать контроллер в ArticlesController

SimeonIksanov commented on 01. Presentation/NewHabr.WebApi/Controllers/ArticleController.cs:
```diff
@@ -0,0 +1,155 @@
+    [HttpGet]
+    public async Task<ActionResult<IEnumerable<ArticleDto>>> GetByTitleAsync(
+        [FromQuery] string title,
+        CancellationToken cancellationToken = default)
```
cancellationToken = default - default тут лишнее, DI всегда подставит сюда токен

SimeonIksanov commented on 01. Presentation/NewHabr.WebApi/Controllers/ArticleController.cs:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    [HttpGet]
+    public async Task<ActionResult<IEnumerable<ArticleDto>>> GetByTitleAsync(
```
суффикс Async у контроллеров не надо писать, кто-то из преподов говорил что это иногда провоцирует какие-то косяки, только не помню какие

SimeonIksanov commented on 01. Presentation/NewHabr.WebApi/Controllers/ArticleController.cs:
```diff
@@ -0,0 +1,155 @@
+    [HttpGet("{userId}")]
+    public async Task<ActionResult<IEnumerable<ArticleDto>>> GetByUserIdAsync(
+    [FromRoute] Guid userId,
+    CancellationToken cancellationToken = default)
```
тут маршрут будет выглядеть как /article/10 - как уже писал, надо изменить на articles/10
и учитывая REST стандарт этот путь должен вернуть статью с ид 10

чтобы получить статьи автора с ид 10 должнен быть путь /users/10/articles

SimeonIksanov commented on 01. Presentation/NewHabr.WebApi/Controllers/ArticleController.cs:
```diff
@@ -0,0 +1,155 @@
+
+    [HttpGet("published")]
+    public async Task<ActionResult<IEnumerable<ArticleDto>>> GetPublishedAsync(CancellationToken cancellationToken = default)
+    {
```
изменить атрибут на unpublished и возвращать неопубликованные статью, потому что опубликованные должны возвращаться без доп приседаний

SimeonIksanov commented on 01. Presentation/NewHabr.WebApi/Controllers/ArticleController.cs:
```diff
@@ -0,0 +1,155 @@
+    {
+        if (string.IsNullOrEmpty(title))
+        {
+            return BadRequest();
```
тут маршрут будет /article и если не указать ?title=blah будет бадреквест??? так быть не должно

SimeonIksanov commented on 01. Presentation/NewHabr.WebApi/Controllers/ArticleController.cs:
```diff
@@ -0,0 +1,155 @@
+    }
+
+    [HttpPut]
+    public async Task<ActionResult> UpdateAsync([FromQuery] ArticleDto articleToUpdate, CancellationToken cancellationToken = default)
```
fromQuery - заменить на fromBody
строка запроса имеет ограничения на длину

SimeonIksanov commented on 01. Presentation/NewHabr.WebApi/Controllers/CategoryController.cs:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    [HttpPost("{name}")]
+    public async Task<ActionResult> CreateAsync([FromRoute] string name, CancellationToken cancellationToken = default)
```
FromRoute --> FromBody

SimeonIksanov commented on 01. Presentation/NewHabr.WebApi/Controllers/CategoryController.cs:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    [HttpPut]
+    public async Task<ActionResult> UpdateAsync([FromQuery] CategoryDto categoryToUpdate, CancellationToken cancellationToken = default)
```
FromRoute --> FromBody

SimeonIksanov replied:
FromQuery --> fromBody

SimeonIksanov commented on 01. Presentation/NewHabr.WebApi/Program.cs:
```diff
@@ -21,6 +26,15 @@ public static void Main(string[] args)
         services.AddSwaggerGen();
 
+        var mapperConfigurations = new MapperConfiguration(config => config.AddMaps(typeof(ArticleProfile).Assembly));
+        var mapper = mapperConfigurations.CreateMapper();
```
Конфигурацию маппера вынести в экстеншен класс, чтобы в program.cs была только одна строка вида
services.ConfigureAutoMapper()

SimeonIksanov commented on 01. Presentation/NewHabr.WebApi/Program.cs:
```diff
@@ -21,6 +26,15 @@ public static void Main(string[] args)
 
+        var mapperConfigurations = new MapperConfiguration(config => config.AddMaps(typeof(ArticleProfile).Assembly));
+        var mapper = mapperConfigurations.CreateMapper();
+        services.AddSingleton(mapper);
```
Конфигурацию маппера вынести в экстеншен класс, чтобы в program.cs была только одна строка вида
services.ConfigureAutoMapper()

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/ArticleService.cs:
```diff
@@ -0,0 +1,105 @@
+
+        var articlesDto = _mapper.Map<List<ArticleDto>>(articles);
+
+        ArgumentNullException.ThrowIfNull(articlesDto, nameof(articlesDto));
```
article это не аргумент
articlesDto это не аргумент

если в базе данных не нашлось статей по фильтру, надо просто вернуть пустую коллекцию

SimeonIksanov commented on 02. Business/NewHabr.Domain/Contracts/ITagRepository.cs:
```diff
@@ -0,0 +1,19 @@
+    /// <exception cref="ArgumentException"></exception>
+    /// <exception cref="ArgumentNullException"></exception>
+    /// <exception cref="OperationCanceledException"></exception>
+    Task DeleteByNameAsync(string name, CancellationToken cancellationToken = default);
```
удалять из базы данных надо по ключу, удаляя так по имени можно вальнуть сразу несколько

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/TagService.cs:
```diff
@@ -0,0 +1,66 @@
+    {
+        ArgumentNullException.ThrowIfNull(name, nameof(name));
+
+        _repositoryManager.TagRepository.Create(new()
```
а если такой тег уже есть?

EvgeniyRuban replied:
Наверное тогда нужен метод AddOrUpdate() для таких сущностей

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/ArticleService.cs:
```diff
@@ -0,0 +1,105 @@
+        ArgumentNullException.ThrowIfNull(article, nameof(article));
+
+        _repositoryManager.ArticleRepository.Update(article);
+        await _repositoryManager.SaveAsync(cancellationToken);
```
Сервис должен сделать запрос к БД и выяснить, если ли статья с таким Id.
если нет  - сбрасывать 404
если есть  - на объект из базы смапить дто из запроса и послать сущность бд в метод репозитория для обновления

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/ArticleService.cs:
```diff
@@ -0,0 +1,105 @@
+    public async Task DeleteByIdAsync(Guid id, CancellationToken cancellationToken = default)
+    {
+        await _repositoryManager.ArticleRepository.DeleteByIdAsync(id, cancellationToken);
+        await _repositoryManager.SaveAsync(cancellationToken);
```
Сервис должен сделать запрос к БД и выяснить, если ли статья с таким Id.
если нет  - сбрасывать 404
если есть  - на объект из базы смапить дто из запроса и послать сущность бд в метод репозитория для удаления

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/ArticleService.cs:
```diff
@@ -0,0 +1,105 @@
+
+        var articles = await _repositoryManager.ArticleRepository.GetByTitleAsync(title, cancellationToken);
+
+        ArgumentNullException.ThrowIfNull(articles, nameof(articles));
```
метод GetByTitleAsync возвращает коллекцию или вываливается по эксепшену(в текущей реализации). В каком случае он вернет null?

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/ArticleService.cs:
```diff
@@ -0,0 +1,105 @@
+
+        var articlesDto = _mapper.Map<List<ArticleDto>>(articles);
+
+        ArgumentNullException.ThrowIfNull(articlesDto, nameof(articlesDto));
```
для чего все эти проверки? не очевидно

EvgeniyRuban replied:
По тому что метод возвращает значение не nullable, по этому я и не стал возвращать null.

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/CategoryService.cs:
```diff
@@ -0,0 +1,66 @@
+        {
+            Name = name,
+        });
+        await _repositoryManager.SaveAsync(cancellationToken);
```
что будет если в бд уже есть такая категория?

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/CategoryService.cs:
```diff
@@ -0,0 +1,66 @@
+    public async Task DeleteByIdAsync(int id, CancellationToken cancellationToken = default)
+    {
+        await _repositoryManager.CategoryRepository.DeleteByIdAsync(id, cancellationToken);
+        await _repositoryManager.SaveAsync(cancellationToken);
```
если там такой нет?

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/CategoryService.cs:
```diff
@@ -0,0 +1,66 @@
+    public async Task DeleteByNameAsync(string name, CancellationToken cancellationToken = default)
+    {
+        await _repositoryManager.CategoryRepository.DeleteByNameAsync(name, cancellationToken);
+        await _repositoryManager.SaveAsync(cancellationToken);
```
по имени не удаляем

EvgeniyRuban committed:
Refactoring

EvgeniyRuban committed:
Refactoring

EvgeniyRuban committed:
Refactoring

EvgeniyRuban committed:
Refactoring

EvgeniyRuban committed:
Merge branch 'articles-businesslogic-impl' of https://github.com/GBDev1-FirstTeam/new-habr into articles-businesslogic-impl

EvgeniyRuban committed:
Refactoring

EvgeniyRuban commented:
Я предлагаю вынести задачу фильтрации получения данных по статьям отдельно, от этой.

EvgeniyRuban committed:
Refactoring

EvgeniyRuban committed:
Refactoring

EvgeniyRuban committed:
Resolved conflict

EvgeniyRuban commented:
При добавлении в таблицу категорий нового имени, которое там уже есть, но помечено как удаленное, происходит ошибка, потому, что свойство Name должно быть уникальным. В связи с этим, я думаю, что операция удаления для сущностей с уникальным свойством (в данном случае Category и Tag), должно быть полной. Из этого вытекает вопрос - "Как поступить с иерархией наследования для сущностей, чтобы избежать зависимости от свойства Deleted?".

Да и при обновлении сущности разумеется та же история!

Нужны предложения.

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/ArticleService.cs:
```diff
@@ -0,0 +1,77 @@
+            throw new ArticleNotFoundException();
+        }
+
+        article = _mapper.Map<Article>(articleToUpdate);
```
так вообще будет работать? здесь маппер создаст новый объект из дто, которого нет в бд

мы ж с тобой не так обговаривали. надо достать статью из бд по ид с трекингом
смапить дто на объект из бд
обновить объект из бд

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/ArticleService.cs:
```diff
@@ -0,0 +1,77 @@
+            throw new ArticleNotFoundException();
+        }
+
+        _repositoryManager.ArticleRepository.Delete(article);
```
для удаления тоже надо искать объект с трекингом

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/CategoryService.cs:
```diff
@@ -0,0 +1,75 @@
+        }
+
+        _repositoryManager.CategoryRepository.Delete(category);
+        await _repositoryManager.SaveAsync(cancellationToken);
```
Возможно что при удалении категории, тегов и прочих , нужно сначала убрать связи между статьями и удаляемой категорией. а то получится что в таблице связей статья-категория, будут мертвые связи

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/CategoryService.cs:
```diff
@@ -0,0 +1,75 @@
+
+    public async Task UpdateAsync(CategoryDto categoryToUpdate, CancellationToken cancellationToken = default)
+    {
+        var targetCategory = await _repositoryManager.CategoryRepository.GetByIdAsync(categoryToUpdate.Id, cancellationToken);
```
трекинг не включен

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/TagService.cs:
```diff
@@ -0,0 +1,75 @@
+
+    public async Task CreateAsync(CreateTagRequest request, CancellationToken cancellationToken = default)
+    {
+        var tag = _repositoryManager.TagRepository.FindByCondition(c => c.Name == request.Name && !c.Deleted).FirstOrDefault();
```
а если в бд есть удаленный тег, мы создадим еще один такой же?

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/TagService.cs:
```diff
@@ -0,0 +1,75 @@
+
+    public async Task DeleteByIdAsync(int id, CancellationToken cancellationToken = default)
+    {
+        var tag = await _repositoryManager.TagRepository.GetByIdAsync(id, cancellationToken);
```
трекинг не включен

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/TagService.cs:
```diff
@@ -0,0 +1,75 @@
+
+    public async Task UpdateAsync(TagDto tagToUpdate, CancellationToken cancellationToken = default)
+    {
+        var targetTag = await _repositoryManager.TagRepository.GetByIdAsync(tagToUpdate.Id, cancellationToken);
```
трекинг не включен

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/TagService.cs:
```diff
@@ -0,0 +1,75 @@
+        }
+
+        var duplicateTag = _repositoryManager.TagRepository
+            .FindByCondition(c => c.Name == tagToUpdate.Name && !c.Deleted).FirstOrDefault();
```
если новое имя тега есть среди удаленных?

EvgeniyRuban replied:
Дак я вроде писал об этом и в телеграмме и здесь. Ты предложил затереть связи, я не понимаю что ты имеешь ввиду под этим. И как это решит проблему добавления в таблицу новых имен категорий, которые там уже есть.

SimeonIksanov replied:
У нас статьи и теги связываются через таблицу связей
если просто пометить тег как удаленный, от этого связи между статьей и тегом не пропадут.
я предлагаю, при "удалении" тега - убирать из таблицы связей связи всех статьей с удаляемым тегом

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/ArticleService.cs:
```diff
@@ -0,0 +1,77 @@
+        return _mapper.Map<List<ArticleDto>>(articles);
+    }
+
+    public async Task CreateAsync(CreateArticleRequest request, CancellationToken cancellationToken = default)
```
при создании статьи, сервис должен проставлять дату создания статьи и дату модификации статьи

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/ArticleService.cs:
```diff
@@ -0,0 +1,77 @@
+        await _repositoryManager.SaveAsync(cancellationToken);
+    }
+
+    public async Task UpdateAsync(ArticleDto articleToUpdate, CancellationToken cancellationToken = default)
```
при обновлении статьи, сервис должен менять дату модификации статьи

SimeonIksanov commented on 02. Business/NewHabr.Business/Services/ArticleService.cs:
```diff
@@ -0,0 +1,77 @@
+
+namespace NewHabr.Business.Services;
+
+public class ArticleService : IArticleService
```
не увидел в сервисе метода для публикации статьи и снятия с публикации

SimeonIksanov commented on 02. Business/NewHabr.Domain/Dto/CreateArticleRequest.cs:
```diff
@@ -0,0 +1,17 @@
+
+    [NotNull]
+    public string Content { get; set; }
+}
```
а категории, теги будут выставлены отдельно от создания?

Commit:
Merge branch 'dev' into articles-businesslogic-impl

# Conflicts:
#	01. Presentation/NewHabr.WebApi/Extensions/ServiceExtensions.cs
#	01. Presentation/NewHabr.WebApi/Program.cs
#	02. Business/NewHabr.Business/AutoMapperProfiles/ArticleProfile.cs
#	02. Business/NewHabr.Business/NewHabr.Business.csproj

EvgeniyRuban committed:
ArticleService logic updated.

EvgeniyRuban committed:
Article mapping profile updated.

EvgeniyRuban committed:
Article mapping profile updated.

EvgeniyRuban committed:
Merge conflict.

EvgeniyRuban committed:
Added logging for controllers.

- ArticleController
- CategoryController
- TagController

EvgeniyRuban committed:
Refactoring

SimeonIksanov commented on 01. Presentation/NewHabr.WebApi/Controllers/ArticlesController.cs:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    [HttpPut("publish/{id}")]
+    public async Task<ActionResult> Publish([FromRoute] Guid id, CancellationToken cancellationToken)
```
путь надо /articles/{id}/publish

SimeonIksanov commented on 01. Presentation/NewHabr.WebApi/Controllers/ArticlesController.cs:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    [HttpPut("unpublish/{id}")]
+    public async Task<ActionResult> Unpublish([FromRoute] Guid id, CancellationToken cancellationToken)
```
/articles/{id}/unpublish

SimeonIksanov commented on 01. Presentation/NewHabr.WebApi/Controllers/ArticlesController.cs:
```diff
@@ -0,0 +1,162 @@
+[Route("api/[controller]")]
+[ApiController]
+public class ArticlesController : ControllerBase
+{
```
в контроллер статей добавить выгрузку комментариев к статье по пути /aritcles/{id}/comments
возможно понадобится выгрузка статей по категории /articles?category="name"
статьи по тегам /articles?tag="name"

EvgeniyRuban committed:
Added new operations.

- set article approve state
- set article  publication status
- update article operation refactoring

EvgeniyRuban committed:
Merge branch 'dev' into articles-businesslogic-impl
